### PR TITLE
[SPARK-29438][SS][FOLLOWUP] Add regression tests for Streaming Aggregation and flatMapGroupsWithState

### PR DESCRIPTION
**Pull request by HeartSaVioR**
### What changes were proposed in this pull request?

This patch adds new UTs to prevent SPARK-29438 for streaming aggregation as well as flatMapGroupsWithState, as we agree about the review comment quote here:

https://github.com/apache/spark/pull/26162#issuecomment-576929692

> LGTM for this PR. But on a additional note, this is a very subtle and easy-to-make bug with TaskContext.getPartitionId. I wonder if this bug is present in any other stateful operation. I wonder if this bug is present in any other stateful operation. Can you please verify how partitionId is used in the other stateful operations?

For now they're not broken, but even better if we have UTs to prevent the case for the future.

### Why are the changes needed?

New UTs will prevent streaming aggregation and flatMapGroupsWithState to be broken in future where it is placed on the right side of UNION and the number of partition is changing on the left side of UNION. Please refer SPARK-29438 for more details.

### Does this PR introduce any user-facing change?

No.

### How was this patch tested?

Added UTs.

### CODE REVIEW

**Comment by HeartSaVioR** [comment]
47351d5 is the only effective commit. Once #26162 is merged I'll rebase this PR.

**Comment by HeartSaVioR** [comment]
cc. @tdas This is a follow-up PR of SPARK-29438 which only adds regression tests. Please take a look. Thanks in advance!

File: sql/core/src/test/scala/org/apache/spark/sql/streaming/FlatMapGroupsWithStateSuite.scala
Comment (tdas) [inline sql/core/src/test/scala/org/apache/spark/sql/streaming/FlatMapGroupsWithStateSuite.scala]: this test does not need these two checks. other tests already test this.

File: sql/core/src/test/scala/org/apache/spark/sql/streaming/FlatMapGroupsWithStateSuite.scala
Comment (tdas) [inline sql/core/src/test/scala/org/apache/spark/sql/streaming/FlatMapGroupsWithStateSuite.scala]: Again I would simplify this test further by having just one batch before and after partition count change. we dont need to validate the whether the state function is being called correctly.

File: sql/core/src/test/scala/org/apache/spark/sql/streaming/FlatMapGroupsWithStateSuite.scala
Comment (tdas) [inline sql/core/src/test/scala/org/apache/spark/sql/streaming/FlatMapGroupsWithStateSuite.scala]: input1 is just unioned at the end and does not need to share the same keys as input2, i would make the input of input1 very different from input2, such that it is obvious in the CheckAnswer which output came from where. For example input1's strings could "input1-a" and input2's can be "input2-a"

File: sql/core/src/test/scala/org/apache/spark/sql/streaming/StreamingAggregationSuite.scala
Comment (tdas) [inline sql/core/src/test/scala/org/apache/spark/sql/streaming/StreamingAggregationSuite.scala]: why do you need 2 column grouping to test this? why not use simply 1 column.

File: sql/core/src/test/scala/org/apache/spark/sql/streaming/StreamingAggregationSuite.scala
Comment (tdas) [inline sql/core/src/test/scala/org/apache/spark/sql/streaming/StreamingAggregationSuite.scala]: same comment as above, make the input space of the two inputs distinct so that it's more obvious which output came from where.

**Comment by HeartSaVioR** [comment]
@tdas 
Thanks for reviewing. Simplified the new UTs as well as differentiate input values between two streams. Please take a look again. Thanks in advance!

**Comment by tdas** [comment]
As is the test look good. I would just test one more thing. If possible, introduce the same bug in aggregate and mapGroupsWithState and see whether these tests fail. You may be able to introduce this by inject the incorrect partition id in the StateStoreRDD when getting the state store.

**Comment by tdas** [comment]
Also for completeness sake, we should add a test for streaming deduplication. That is the most like operation that people might do with a union.

**Comment by HeartSaVioR** [comment]
> If possible, introduce the same bug in aggregate and mapGroupsWithState and see whether these tests fail. You may be able to introduce this by inject the incorrect partition id in the StateStoreRDD when getting the state store.

Confirmed below code change makes new tests fail for same reason of SPARK-29438:

```
diff --git a/sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/StateStoreRDD.scala b/sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/StateStoreRDD.scala
index 4a69a48fed..6e58017a64 100644
--- a/sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/StateStoreRDD.scala
+++ b/sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/StateStoreRDD.scala
@@ -69,9 +69,6 @@ class StateStoreRDD[T: ClassTag, U: ClassTag](

   override def compute(partition: Partition, ctxt: TaskContext): Iterator[U] = {
     var store: StateStore = null
-    val storeProviderId = StateStoreProviderId(
-      StateStoreId(checkpointLocation, operatorId, partition.index),
-      queryRunId)

     // If we're in continuous processing mode, we should get the store version for the current
     // epoch rather than the one at planning time.
@@ -85,6 +82,16 @@ class StateStoreRDD[T: ClassTag, U: ClassTag](
       storeVersion
     }

+    val newPartIdx = if (currentVersion > 0) {
+      2 + partition.index
+    } else {
+      partition.index
+    }
+
+    val storeProviderId = StateStoreProviderId(
+      StateStoreId(checkpointLocation, operatorId, newPartIdx),
+      queryRunId)
+
     store = StateStore.get(
       storeProviderId, keySchema, valueSchema, indexOrdinal, currentVersion,
       storeConf, hadoopConfBroadcast.value.value)
```

> Also for completeness sake, we should add a test for streaming deduplication.

OK I'll add it sooner. Thanks!

**Comment by HeartSaVioR** [comment]
Also added UT for streaming deduplication as well. For sure, the UT fails under code modification as well in above comment.

**Comment by HeartSaVioR** [comment]
cc. @tdas Kindly reminder.

**Comment by HeartSaVioR** [comment]
cc. @tdas @zsxwing ping.

**Comment by HeartSaVioR** [comment]
cc. @tdas Kindly reminder.

**Comment by HeartSaVioR** [comment]
retest this, please

**Comment by HeartSaVioR** [comment]
retest this, please

**Comment by HeartSaVioR** [comment]
retest this, please

**Comment by HeartSaVioR** [comment]
retest this, please

**Comment by HeartSaVioR** [comment]
retest this, please

**Comment by HeartSaVioR** [comment]
retest this, please

**Comment by HeartSaVioR** [comment]
retest this, please

**Comment by HeartSaVioR** [comment]
retest this, please

**Comment by HeartSaVioR** [comment]
retest this, please

**Comment by HeartSaVioR** [comment]
retest this, please

**Comment by HeartSaVioR** [comment]
retest this, please

**Comment by HeartSaVioR** [comment]
cc. @tdas @zsxwing @gaborgsomogyi @viirya @xuanyuanking

**Comment by HeartSaVioR** [comment]
retest this, please

File: sql/core/src/test/scala/org/apache/spark/sql/streaming/FlatMapGroupsWithStateSuite.scala
Comment (viirya) [inline sql/core/src/test/scala/org/apache/spark/sql/streaming/FlatMapGroupsWithStateSuite.scala:1062]: Why we need to add data here? Don't we rely on `MultiAddData` below to add more data to `newInput1` and `newInput2`?

**Review by viirya — COMMENTED** [review]
This overall looks good to have. Just one question.

**Comment by HeartSaVioR** [comment]
Accidentally closed this - don't get the UX on Github mobile app. (Don't know how to reopen in the app.) Will reopen soon.

File: sql/core/src/test/scala/org/apache/spark/sql/streaming/FlatMapGroupsWithStateSuite.scala
Comment (HeartSaVioR) [reply sql/core/src/test/scala/org/apache/spark/sql/streaming/FlatMapGroupsWithStateSuite.scala:1062]: We're initializing the different memory streams here, which are not aware of previous data and require us to fill the same data again on restore. (Each addition to the memory stream represents a batch, hence you'll need to add the same for the new memory stream.)

Probably there could be the another kind of small improvement for tests if we have convenient way to push all the input rows we put before to the new memory stream. That may be dealt with minor PR.

**Review by viirya — APPROVED** [review]

**Comment by viirya** [comment]
retest this please

**Comment by viirya** [comment]
Will merge this once jenkins pass.

**Comment by viirya** [comment]
Thanks. Merging to master.

**Comment by HeartSaVioR** [comment]
Thanks all for reviewing and merging!

**Pull request closed without merging**
